[TW-1649] Postman Vault February release items (#5396)
postmanlabs/postman-docs: src/pages/docs/sending-requests/postman-vault-secrets.md

* [Access your Postman Vault](#access-your-postman-vault)
* [Manage your vault key](#manage-your-vault-key)
    * [Save your vault key](#save-your-vault-key)
    * [Reset your vault key](#reset-your-vault-key)
* [Add sensitive data as vault secrets](#add-sensitive-data-as-vault-secrets)
* [Edit vault secrets](#edit-vault-secrets)
* [Use vault secrets](#use-vault-secrets)
* [Fix unresolved vault secrets](#fix-unresolved-vault-secrets)
* [Web and desktop limitations](#web-and-desktop-limitations)

## About vault secrets

*Vault secrets* are sensitive data, such as API keys and passwords, that you store in your Postman Vault and reuse in your local instance of Postman. Only you can access and reuse your vault secrets in your local instance of Postman, and they aren't synced to the Postman cloud.

Vault secrets are useful when:

* You're using the same sensitive data in multiple places throughout your workspaces.
* You don't want your sensitive data synced to the Postman cloud.
* You want to be the only user with access to your sensitive data.
* You want to specify the domains and subdomains you're allowed to send your sensitive data to.

> * You want your sensitive data encrypted, requiring a unique key to access it.

Reviewer: The vault key is not the actual encryption key, and we do not intend for user to manage their vault encryption key may be we can remove the second part and simply say,
> You want your sensitive data encrypted.

Author: This is now updated. Please see https://github.com/postmanlabs/postman-docs/pull/5396/commits/6a0165940a3bbf18ee24a743bad41e7a31711abb

> If a variable in a different scope has the same name as a vault secret, the value stored in a narrower scope won't be used. For example, if there is a collection variable named `postman-api-key` and a vault secret also named `postman-api-key`, the collection value won't be used when the request runs. Learn how to [reference vault secrets](#use-vault-secrets) in Postman.

### Compare vault secrets and variables

Both vault secrets and variables enable you to store and reference data in Postman. One may be a better fit than the other, depending on your use case or preferences.

[Vault secrets](#about-vault-secrets) enable you to store and reuse sensitive data only in your local instance of Postman, and they aren't synced to the Postman cloud. This enables you to keep your sensitive data hidden from collaborators, and prevent collaborators from using your vault secrets. Vault secrets are masked by default, and only you can unmask them. Your sensitive data is also encrypted in your Postman Vault, requiring your vault key to gain access.

> [Variables](/docs/sending-requests/variables/) enable you to store and reuse the same value, such as URLs or API keys, in multiple places. The following are some options for storing data in variables:

Reviewer: We also mentioned "API keys" in this statement, which might confuse the users regarding where to store them. After introducing Postman Vault, we want all Postman users to store sensitive data, including credentials, inside the vault.

Author: This is now updated. Please see https://github.com/postmanlabs/postman-docs/pull/5396/commits/6a0165940a3bbf18ee24a743bad41e7a31711abb

* You want to be the only user with access to your sensitive data.
* You want to specify the domains and subdomains you're allowed to send your sensitive data to.
* You want your sensitive data encrypted, requiring a unique key to access it.

Vault secrets are created at the *vault* scope, which is the broadest scope compared to the [variable scopes](/docs/sending-requests/variables/#variable-scopes) in Postman. This enables you to securely access vault secrets at the team level, and reuse them in your collections, requests, and environments throughout your workspaces.

> If a variable in a different scope has the same name as a vault secret, the value stored in a narrower scope won't be used. For example, if there is a collection variable named `postman-api-key` and a vault secret also named `postman-api-key`, the collection value won't be used when the request runs. Learn how to [reference vault secrets](#use-vault-secrets) in Postman.

### Compare vault secrets and variables

Both vault secrets and variables enable you to store and reference data in Postman. One may be a better fit than the other, depending on your use case or preferences.

> [Vault secrets](#about-vault-secrets) enable you to store and reuse sensitive data only in your local instance of Postman, and it isn't synced to the Postman cloud. This enables you to keep your sensitive data hidden from collaborators, and prevent collaborators from using your vault secrets. Vault secrets are masked by default, and only you can unmask them. Your sensitive data is also encrypted in your Postman Vault, requiring your vault key to gain access.

Reviewer 2: ```suggestion
[Vault secrets](#about-vault-secrets) enable you to store and reuse sensitive data only in your local instance of Postman, and they aren't synced to the Postman cloud. This enables you to keep your sensitive data hidden from collaborators, and prevent collaborators from using your vault secrets. Vault secrets are masked by default, and only you can unmask them. Your sensitive data is also encrypted in your Postman Vault, requiring your vault key to gain access.
```

> [Variables](/docs/sending-requests/variables/) enable you to store and reuse the same value, such as URLs or API keys, in multiple places. The following includes some options for storing data in variables:

Reviewer 2: ```suggestion
[Variables](/docs/sending-requests/variables/) enable you to store and reuse the same value, such as URLs or API keys, in multiple places. The following are some options for storing data in variables:
```

## Access your Postman Vault

To access your Postman Vault, open a workspace then select <img alt="Vault icon" src="https://assets.postman.com/postman-docs/icons/icon-postman-vault.jpg#icon" width="14px"> **Vault** from the Postman footer. You can also use **Control+Shift+V** or **Ctrl+Shift+V** to access your Postman Vault.

You can open your vault secrets in Postman Vault as follows:

* If this is your first time opening your Postman Vault, select **Set Up Vault** to [get your vault key](#manage-your-vault-key).
* If Postman already generated your [vault key](#manage-your-vault-key), enter your vault key then select **Open Vault**. Then you can continue to [add](#add-sensitive-data-as-vault-secrets), [edit](#edit-vault-secrets), and [use](#use-vault-secrets) your vault secrets in Postman.

>     > If you stored your vault key in your system's password manager, Postman automatically gets your vault key. You must use the [Postman Desktop Agent](/docs/getting-started/basics/about-postman-agent/#the-postman-desktop-agent).

Reviewer 2: Mention that the Desktop Agent is just required if you're using the web app

Author: Good idea. Thanks!

* You can set the variable type as [secret type](/docs/sending-requests/variables/#variable-types) in global and environment variables. This enables you to mask sensitive data in the initial and current values. Collaborators with access to the workspace can [view a secret type variable's values](/docs/sending-requests/variables/#viewing-and-changing-secret-variable-values). Also, collaborators with additional permissions can [change the variable type](/docs/sending-requests/variables/#changing-from-secret-to-default-variable-type), unmasking it for collaborators.

### Web and desktop limitations

Some Postman Vault features require the Postman desktop app or the Postman web app.

The following features require the Postman desktop app:

* **Open Postman Vault from public workspaces** - You must use the Postman desktop app to open your Postman Vault from a [public workspace](/docs/collaborating-in-postman/using-workspaces/public-workspaces/), and reference vault secrets in a public workspace. If you're using the Postman web app, you must add new vault secrets to your Postman Vault if you're opening it from a public workspace.

> * **Use vault secrets between teams** - You must use the Postman desktop app to continue using your vault secrets when you [join, switch, or leave teams](/docs/collaborating-in-postman/working-with-your-team/team-collaboration/). If you're using the Postman web app, you must add new vault secrets to your Postman Vault in your other teams.

Reviewer 3: `If you're using the Postman web app, you must add new vault secrets to your Postman Vault in your other teams`
This gives me the idea that I need to add new ones in web app but old secrets are retained with desktop app. That is the case with team joining and team leaving. When you switch between teams, secrets are never retained.

Author: This is now updated. Please see the updated wording at https://github.com/postmanlabs/postman-docs/pull/5396/commits/c105b4621ab88e01e053129ecd732611653aff11

## Access your Postman Vault

To access your Postman Vault, open a workspace then select <img alt="Vault icon" src="https://assets.postman.com/postman-docs/icons/icon-postman-vault.jpg#icon" width="14px"> **Vault** from the Postman footer. You can also use **Control+Shift+V** or **Ctrl+Shift+V** to access your Postman Vault.

You can open your vault secrets in Postman Vault as follows:

* If this is your first time opening your Postman Vault, select **Set Up Vault** to [get your vault key](#manage-your-vault-key).
* If Postman already generated your [vault key](#manage-your-vault-key), enter your vault key then select **Open Vault**. Then you can continue to [add](#add-sensitive-data-as-vault-secrets), [edit](#edit-vault-secrets), and [use](#use-vault-secrets) your vault secrets in Postman.

>     > If you stored your vault key in your system's password manager, Postman automatically gets your vault key. You must use the [Postman Desktop Agent](/docs/getting-started/basics/about-postman-agent/#the-postman-desktop-agent) if you're using the the [Postman web app](/docs/getting-started/installation/installation-and-updates/#use-the-postman-web-app) so Postman can get your vault key.

Reviewer 3: @MatthewPinsker-postman we can provide an example here for the system's password manager like Keychain Access for MacOS

>     > If you stored your vault key in your system's password manager, Postman automatically gets your vault key. You must use the [Postman Desktop Agent](/docs/getting-started/basics/about-postman-agent/#the-postman-desktop-agent) if you're using the the [Postman web app](/docs/getting-started/installation/installation-and-updates/#use-the-postman-web-app) so Postman can get your vault key.

Reviewer 3: @MatthewPinsker-postman I think we can combine this message and the one in `Manage your vault key` section into one as they both convey the same information

Author: Good question! Thanks for bringing this up.

I'll remove the note from the "Manage your vault key" section since we also mention this note in it's subsection, "Save your vault key". This information is more relevant to the procedural task.

For "Access your Postman Vault" and "Save your vault key" sections, I think it's helpful to mention in both places because we're asking users to complete different tasks, and we expect users to skip around the page.

Your vault secrets are encrypted using the AES-256-GCM encryption algorithm, and your vault key isn't synced to the Postman cloud.

When Postman generates your vault key, you have the following options for saving your key:

* Save your vault key in a secure location. Each time you sign in to Postman, you'll need to manually enter your vault key, enabling you to access your Postman Vault and vault secrets.
* Postman can store your vault key in your system's password manager, such as Keychain Access on macOS. This enables Postman to automatically get your vault key, giving you access to your Postman Vault and vault secrets when you sign in to Postman.

    > If you're using Postman Vault from the [Postman web app](/docs/getting-started/installation/installation-and-updates/#use-the-postman-web-app), you must use the [Postman Desktop Agent](/docs/getting-started/basics/about-postman-agent/#the-postman-desktop-agent) so Postman can get your vault key.

### Save your vault key

> 1. Open your Postman Vault. If this is your first time opening your Postman Vault, select **Encrypt Vault**.

Reviewer 3: @MatthewPinsker-postman When the user opens it for the first time, they get to see `Set up Vault`. Encrypt Vault is rather a flow which existing users will go through as part of data migration which needn't necessarily be convered under LC doc

Author: Thanks for clarifying! This is now updated.

When you open Postman Vault for the first time, Postman generates a [vault key](#save-your-vault-key) that's used to access your encrypted vault secrets. You can [reset your vault key](#reset-your-vault-key) if you lose it, but Postman will permanently delete your vault secrets. Once you save your vault key, you can [add vault secrets to your Postman Vault](#add-sensitive-data-as-vault-secrets).

> Your vault secrets are encrypted using the AES-256-GCM encryption algorithm. Also your vault key isn't synced to the Postman cloud.

Reviewer: @MatthewPinsker-postman let's say "...your encryption key isn't synced to the Postman cloud."  

Technical detail just for your understanding, no need to mention in the LC:
We hash the vault key and send it to the Potman cloud for authentication purposes. The value we get after hashing the vault key can never give the original vault key. Hence better to use "encryption key" here which we never sync to Postman cloud. cc @Suhas-Gaikwad

Author: As discussed via Slack, we'll remove this sentence for now. Thanks!

> If you're using the Postman web app to send requests with references to vault secrets, you must use the [Postman Desktop Agent](/docs/getting-started/basics/about-postman-agent/#the-postman-desktop-agent) or the [Postman Browser Agent](/docs/getting-started/basics/about-postman-agent/#the-postman-browser-agent).

Enclose the vault secret in double curly braces (`{{ }}`) and prefix the secret's name with `vault:` to reference it throughout your Postman team. For example, to reference a secret named "postman-api-key", use the following syntax:

```txt
{{vault:postman-api-key}}
```

<!-- ### Use vault secrets in scripts

You can get the current value of a vault secret in your scripts using the following syntax:

```js
//access a vault secret in you Postman Vault
pm.variables.get("vault:variable-key");
``` -->

![Reference vault secrets in Postman](https://assets.postman.com/postman-docs/v10/use-postman-vault-secrets-v10-23.jpg)

> > If you reference a vault secret as the initial value of a variable, the reference to the secret (for example `{{vault:secret-name}}`) is synced using Postman's cloud servers, and shared with anyone who has access to the workspace. The secret stored in your Postman Vault isn't synced or shared. Learn more about [initial and current values](/docs/sending-requests/variables/#initial-and-current-values).

Reviewer: > reference a vault secret as the initial value of a variable
I didn't get this 😅 vault secrets don't have initial values and are not synced to cloud

Author: @ronakodhaviya Thanks for bringing this up! My intention here to let users know that if they reference a vault secret in the initial value of a variable, such as an environment variable, anyone with access to the workspace will see the secret's name. This is for users who might not want their secret's name shared with other Postman users.

Please let me know if this isn't a concern, and I can remove the note.

Author: I discussed this with @Suhas-Gaikwad , and he confirmed that my understanding is correct. I made a few changes to the note to help clarify. Please see https://github.com/postmanlabs/postman-docs/pull/5396/commits/6a0165940a3bbf18ee24a743bad41e7a31711abb

Enclose the vault secret in double curly braces (`{{ }}`) and prefix the secret's name with `vault:` to reference it throughout your Postman team. For example, to reference a secret named "postman-api-key", use the following syntax:

```txt
{{vault:postman-api-key}}
```

<!-- ### Use vault secrets in scripts

You can get the current value of a vault secret in your scripts using the following syntax:

```js
//access a vault secret in you Postman Vault
pm.variables.get("vault:variable-key");
``` -->

![Reference vault secrets in Postman](https://assets.postman.com/postman-docs/v10/use-postman-vault-secrets-v10-23.jpg)

> > If you reference a vault secret as the initial value of a variable, the reference to the secret (for example `{{vault:secret-name}}`) is synced using Postman's cloud servers, and shared with anyone who has access to the workspace. The secret stored in your Postman Vault isn't synced or shared. Learn more about [initial and current values](/docs/sending-requests/variables/#initial-and-current-values).

Reviewer 3: @MatthewPinsker-postman we can also mention that vault secrets are not usable with cloud agents

Author: This is now updated. https://github.com/postmanlabs/postman-docs/pull/5396/commits/77cab13f6d0b775c42a70fc8ee797153c58f9c25

I followed up via Slack with additional questions.
